odoo/models.py

```diff
@@ -6835,7 +6835,8 @@ def _modified_triggers(self, tree, create=False):
                     records = model.search([(field.name, 'in', real_records.ids)], order='id')
                 if new_records:
                     cache_records = self.env.cache.get_records(model, field)
-                    records |= cache_records.filtered(lambda r: set(r[field.name]._ids) & set(self._ids))
+                    new_ids = set(self._ids)
+                    records |= cache_records.filtered(lambda r: not set(r[field.name]._ids).isdisjoint(new_ids))
```

Comment:
`set(r[field.name]._ids)` should never contain duplicates (AFAIK, X2Many is always duplicated in the cache), then the set seems redundant. Also, since isdisjoint() accepts any Iterable, then an alternative solution might be `not new_ids.isdisjoint(r[field.name]._ids)`.

I am really not sure if it is better because of the Python optimization done when we use isdisjoint with 2 set (https://github.com/python/cpython/blob/333c7dccd87c637d0b15cf81f9bbec28e39664fd/Objects/setobject.c#L1351-L1373) but it is worth it to try.

Reply:
> I still have one question: after this change, is the speedscope still showing this part of the code, or it become insignificant ?

The `__hash__` bottleneck completely disappear (a few 20ms calls, basically nothing), the `_modified_trigger` itself is still present, but I'm somewhat expecting that, the trigger tree is large, because `allow_billable` is a frequent dependency, and there is business code that can be fixed, I'm still investigating that.

Left heavy profile post-patch, same instance as PR:
![profile_onchange_billable_after_orm_patch](https://github.com/odoo/odoo/assets/111067497/c9287c31-1127-4972-b5db-11da5f5f522c)

As for your suggestion, I quickly tested/benched it, on hot, the execution time are almost identical (36s for current patch, 38s for `not new_ids.isdisjoint(r[field.name]._ids)`) which could be attributed to a margin of error. (it's different timings than my PR benchmark, because the staging was reset and code updated in the meantime '-.-, so I rebenched it).

I guess building the `set` python side and then hitting the shortcut in CPython is somewhat equivalent in term of performance than leaving the iteration inside CPython. 🤷🏼‍♂️